### PR TITLE
Modular powdr syntax for single machine

### PR DESCRIPTION
**Pull request by Schaeff**
This aims to be a minimal change which introduces a new modular syntax for powdr.

The parser supports trees of machines, however the compiler panics for any tree which is not 
- a single machine 
- with a single operation 
- called main 
- with no inputs and no outputs

In this special case:
- single machine: the linker simply uses the output of compiling the main machine (in the future: use the object tree (compiled machine instances) and instantiate their links using for example lookups)
- single operation: rom generation simply uses the main operation (in the future: dispatch calls to the different operations of the machine)

### CODE REVIEW

File: test_data/asm/batching/labels.asm [hunk 1]
```diff
@@ -1,17 +1,21 @@
...
+    reg pc[@pc];
+    reg X[<=];
+    reg A;
 
-instr loop { pc' = pc }
+    instr loop { pc' = pc }
 
-label_with_next::
- A <=X= 1;
+    operation main {
```
Comment (chriseth) [inline test_data/asm/batching/labels.asm:9]: What is the difference between `operation main` and `program`?

File: ast/src/object/mod.rs [hunk 2]
```diff
@@ -0,0 +1,27 @@
+use std::collections::BTreeMap;
+
+use crate::parsed::PilStatement;
+
+mod display;
+
+#[derive(Default, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Location {
+    limbs: Vec<String>,
```
Comment (leonardoalt) [inline ast/src/object/mod.rs:9]:
```suggestion
    path: Vec<String>,
```
for clarity

File: ast/src/asm_analysis/mod.rs [hunk 3]
```diff
@@ -1,94 +1,149 @@
...
+
+    pub fn write_registers(&self) -> impl Iterator<Item = &RegisterDeclarationStatement> {
+        self.registers.iter().filter(|r| r.flag.is_none())
+    }
+
+    pub fn assignment_registers(&self) -> impl Iterator<Item = &RegisterDeclarationStatement> {
+        self.registers
+            .iter()
+            .filter(|r| r.flag == Some(RegisterFlag::IsAssignment))
+    }
```
Comment (leonardoalt) [inline ast/src/asm_analysis/mod.rs:134]: Maybe remove

File: ast/src/asm_analysis/display.rs [hunk 4]
```diff
@@ -2,36 +2,57 @@ use std::fmt::{Display, Formatter, Result};
...
+            writeln!(f, "\t{s}")?;
         }
         for s in &self.registers {
-            writeln!(f, "{s}")?;
+            writeln!(f, "\t{s}")?;
         }
-        for s in &self.pil {
-            writeln!(f, "{s}")?;
+        for s in &self.constraints {
+            writeln!(f, "\t{s}")?;
```
Comment (leonardoalt) [inline ast/src/asm_analysis/display.rs:31]: Maybe add a TODO comment to do this properly later with scope sizes

File: ast/src/analysis/mod.rs [hunk 5]
```diff
@@ -0,0 +1,106 @@
+mod display;
```
Comment (leonardoalt) [inline ast/src/analysis/mod.rs:1]: This file can maybe be deleted entirely.

**Review by leonardoalt — COMMENTED** [review]
Comments for the `ast` crate.

File: test_data/asm/batching/labels.asm (hunk 1 above)
Comment (Schaeff) [reply test_data/asm/batching/labels.asm:9]: `program` does not exist in the parser anymore. It is the rom generated during compilation by joining all operations.

File: ast/src/asm_analysis/mod.rs [hunk 6]
```diff
@@ -1,94 +1,139 @@
...
+}
+
+#[derive(Clone, Default)]
+pub struct Program<T> {
+    pub statements: Vec<OperationStatement<T>>,
     pub batches: Option<Vec<BatchMetadata>>,
 }
 
+pub struct AnalysisASMFile<T> {
+    pub machines: HashMap<String, Machine<T>>,
```
Comment (leonardoalt) [inline ast/src/asm_analysis/mod.rs:134]: Check this for nondeterminism also.

File: analysis/src/type_check.rs [hunk 7]
```diff
@@ -1,75 +1,186 @@
...
+                .machines_types
+                .drain()
+                .map(|(name, machine)| (name, machine.unwrap()))
+                .collect();
+            Ok(AnalysisASMFile { machines })
+        }
+    }
+}
+
+/// A very stupid type checker. TODO: make it smart
```
Comment (leonardoalt) [inline analysis/src/type_check.rs:183]: Either remove comment or explain what `smart` means

File: analysis/src/romgen.rs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+                .collect(),
+        }
+    }
+
+    fn generate_machine_rom(&self, mut machine: Machine<T>) -> Machine<T> {
+        machine.program = machine.has_pc().then(|| {
+            assert_eq!(
+                machine.operations.len(),
+                1,
+                "only a single operation is supported"
```
Comment (leonardoalt) [inline analysis/src/romgen.rs:32]: Add a comment at the beginning of the file explaining why this romgen is this simple

File: asm_to_pil/src/lib.rs [hunk 9]
```diff
@@ -901,6 +969,30 @@ fn witness_column<S: Into<String>, T>(
...
+//     op: BinaryOperator,
+//     right: Expression<T>,
+// ) -> Expression<T> {
+//     Expression::BinaryOperation(Box::new(left), op, Box::new(right))
+// }
+
+// fn build_number<T: FieldElement, V: Into<T>>(value: V) -> Expression<T> {
+//     Expression::Number(value.into())
+// }
+
```
Comment (leonardoalt) [inline asm_to_pil/src/lib.rs:995]: Do something about this block.

**Review by leonardoalt — COMMENTED** [review]
Next batch of comments.

File: parser/src/powdr.lalrpop [hunk 10]
```diff
@@ -13,17 +13,17 @@ match {
 }
```
Comment (leonardoalt) [inline parser/src/powdr.lalrpop:14]: @chriseth please review this file instead of me

File: linker/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,167 @@
+use ast::{
+    object::{Location, PILGraph},
+    parsed::{Expression, PILFile, PilStatement},
+};
+use number::FieldElement;
+
+/// a monolithic linker which outputs a single circuit
```
Comment (leonardoalt) [inline linker/src/lib.rs:7]:
```suggestion
/// a monolithic linker which outputs a single AIR
```
or something else

File: test_data/asm/batching/labels_batched.asm [hunk 12]
```diff
@@ -1,21 +1,16 @@
-degree 256;
-reg pc[@pc];
-reg X[<=];
-reg A;
+program {
```
Comment (leonardoalt) [inline test_data/asm/batching/labels_batched.asm:1]: Maybe add a comment saying this isn't valid assembly, and/or change the file extension to make it explicitly that this is a specific test for batching.

**Review by leonardoalt — COMMENTED** [review]
Last batch for now.

**Comment by Schaeff** [comment]
Last suggestions I would like to apply in this PR:
- rename operation to function everywhere
- rename program to rom to avoid confusion with the current program

**Comment by leonardoalt** [comment]
Sounds good, maybe do that in one separate commit each so the commit diff is easy to review

**Review by leonardoalt — COMMENTED** [review]
Current version lgtm

**Review by leonardoalt — APPROVED** [review]
lgtm, @chriseth do you wanna take a look or fine like this?

@Schaeff please squash

**Pull request merged**